File: data_structure/stack/stack.py
"""
Stack
"""
import unittest
from typing import TypeVar, Generic, Generator, Any

T = TypeVar('T')


class Stack(Generic[T]):

    def __init__(self, cap: int = 100):
        self._data: list[T] = [None] * cap
        self._cap = cap
        self._top = -1

    def __iter__(self) -> Generator[T, Any, None]:
        if not self.is_empty():
            for i in range(self._top + 1):
                yield self._data[i]

    def __repr__(self) -> str:
        return ','.join([e for e in self])

    @property
    def size(self) -> int:
        return self._top + 1

    def is_empty(self) -> bool:
        return self._top < 0

    def is_full(self) -> bool:
        return self.size == self._cap

    def top_value(self) -> T:
        if self.is_empty():
            raise RuntimeError('Stack underflow, can not get top')
        return self._data[self._top]

    def push(self, value: T):
        if self.is_full():
            raise RuntimeError('Stack overflow, can not push')
        self._top += 1
        self._data[self._top] = value

    def pop(self) -> T:
        if self.is_empty():
            raise RuntimeError('Stack underflow, can not pop')
        removed = self._data[self._top]
        self._top -= 1
        return removed


class TestStack(unittest.TestCase):

    def test_integer_stack(self):
        stack: Stack[int] = Stack(8)

        self.assertTrue(stack.is_empty())
        self.assertEqual(stack.size, 0)

        stack.push(8)
        stack.push(6)
        stack.push(26)
        stack.push(29)
        stack.push(2)
        stack.push(9)
        stack.push(7)
        stack.push(22)

        self.assertFalse(stack.is_empty())
        self.assertListEqual([e for e in stack],
                             [8, 6, 26, 29, 2, 9, 7, 22])

        with self.assertRaises(RuntimeError):
            stack.push(19)

        self.assertEqual(stack.top_value(), 22)
        self.assertEqual(stack.pop(), 22)
        self.assertEqual(stack.top_value(), 7)
        self.assertEqual(stack.pop(), 7)

        self.assertEqual(stack.size, 6)

        self.assertEqual(stack.pop(), 9)
        self.assertEqual(stack.pop(), 2)
        self.assertEqual(stack.pop(), 29)
        self.assertEqual(stack.pop(), 26)

        self.assertListEqual([e for e in stack], [8, 6])

        self.assertEqual(stack.pop(), 6)
        self.assertEqual(stack.pop(), 8)

        with self.assertRaises(RuntimeError):
            stack.pop()


if __name__ == '__main__':
    unittest.main()

File: data_structure/stack/stack.c
#include "stack.h"
#include <assert.h>
#include <stdlib.h>

/**
 * Stack implementation
 */
cstack_t *create_stack(int cap) {
  cstack_t *stack = (cstack_t *)malloc(sizeof(cstack_t));
  stack->data = malloc(sizeof(void *) * cap);
  stack->cap = cap;
  stack->top = -1;
  return stack;
}

void delete_stack(cstack_t *stack) {
  free(stack->data);
  free(stack);
}

void print_stack(cstack_t *stack) {
  for (int i = 0; i <= stack->top; i++) {
    printf("%p ", stack->data[i]);
  }
}

int size(cstack_t *stack) { return stack->top + 1; }

bool is_empty(cstack_t *stack) { return stack->top < 0; }

bool is_full(cstack_t *stack) { return size(stack) == stack->cap; }

bool push(cstack_t *stack, void *data) {
  if (is_full(stack)) {
    return false;
  }

  stack->data[++stack->top] = data;
  return true;
}

void *pop(cstack_t *stack) {
  if (is_empty(stack)) {
    return NULL;
  }

  return stack->data[stack->top--];
}

const void *top(cstack_t *stack) {
  if (is_empty(stack)) {
    return NULL;
  }

  return stack->data[stack->top];
}

// Testing
void test() {
  cstack_t *stack = create_stack(20);
  assert(is_empty(stack));
  assert(size(stack) == 0);

  push(stack, (void *)8);
  push(stack, (void *)6);
  push(stack, (void *)26);
  push(stack, (void *)9);
  push(stack, (void *)7);
  push(stack, (void *)22);

  assert(!is_empty(stack));
  assert(size(stack) == 6);

  assert((int)top(stack) == 22);
  assert((int)pop(stack) == 22);
  assert(size(stack) == 5);

  assert((int)pop(stack) == 7);
  assert((int)top(stack) == 9);
  assert(size(stack) == 4);

  delete_stack(stack);
  printf("PASSED\n");
}

int main(int argc, const char *argv[]) {
  printf("Testing stack implementation...\n");
  test();
  return 0;
}
